oggquiz.1
.\"
.\"  "THE BEER-WARE LICENSE" (Revision 42):
.\"  <tobias.rehbein@web.de> wrote this file. As long as you retain this notice
.\"  you can do whatever you want with this stuff. If we meet some day, and you
.\"  think this stuff is worth it, you can buy me a beer in return.
.\"                                                               Tobias Rehbein
.\"  
.
.Dd March 25, 2009
.Os 
.Dt OGGQUIZ 1
.
.Sh NAME
.
.Nm oggquiz
.Nd a simple quiz on songs from your ogg/vorbis collection
.
.
.Sh SYNOPSIS
.
.Nm
.Op Fl t | Fl -time Ar seconds
.Op Fl c | Fl -choices Ar choices
.Op Fl p | Fl -players Ar players
.Op Fl o | Fl -ogg123-options Ar options
.Op Fl O | Fl -ogg123 Ar command
.Pp
.Nm
.Brq Fl h | Fl -help
.
.
.Sh DESCRIPTION
.
.Nm
is played in turns. Each turn one song is played and the current player is asked
which song he is listening to. This is done in a multiple choice manner.
.Pp
The faster the player answers the less points he gets added to his current
score. If the player gives the wrong answer or waits too long he gets the
maximum amount of possible points added to his score.
.Pp
The player with the least points wins.
.
.
.Sh OPTIONS
.
.Bl -tag
.It Fl t | Fl -time Ar seconds
Sets the time each player has to make his guess. Also determines the maximum
score which equals
.Ar seconds .
The default is 60.
.
.It Fl c | Fl choices Ar choices
Determines how many possible answers are presented to the player. The default
is 4.
.Ar choices
must not exceed 4 and must not be less than 1. The default is 4.
.
.It Fl p | Fl -players Ar players
Determines how many players are playing the game.
.Ar players
must not exceed 4 and must not be less than 1. The default is 4.
.
.It Fl o | Fl -ogg123-options Ar options
The given
.Ar options
are passed on to ogg123.
.
.It Fl O | Fl -ogg123 Ar command
The explicit command used to invoke ogg123. Might be another player as as it
supports ogg/vorbis and provides a
.Fl q 
flag with the same semantics as ogg123's
.FL q 
flag . The default is
.Dq /usr/local/bin/ogg123 .
.
.It Fl h | Fl -help
Print the synopsis of
.Nm
and exit. This overrides any other options.
.
.El
.
.
.Sh HOW IT WORKS
.
.Nm
Reads a randomized list of ogg/vorbis files from stdin. Having read
.Ar choices
(see OPTIONS) files it stops and picks one of the just read songs and plays it
via ogg123. It then presents the current player with a list of the read songs
and asks which one is playing.
.Pp
If the player gives the wrong answer he gets the full
.Ar seconds
(see OPTIONS) added to his score.
.Pp
If the player gives the correct answer and he used less then
.Ar seconds
(see OPTIONS) seconds to give the answer he gets the amount of seconds he used
added to his score. If the player uses more than
.Ar seconds
(see OPTIONS) seconds to give the answer he gets the same points as if he had
given the wrong answer.
.Pp
After each turn a scoreboard is displayed. Always remember that the player with
the lowest score wins.
.Pp
.Nm 
has no real end. It ends if no more songs can be read from stdin or if one of
the players decides to press the 'q' button and quit the game.
.
.
.Sh ENVIRONMENT
.
.Nm
peruses the
.Ev LANG
and
.Ev LC_ALL
environment variables to determine the systems locale.
.
.Sh EXAMPLES
.
I start my games like this:
.Pp
.Dl find /my/music -type f -name '*.ogg' | random -l -f - | oggquiz
.
.
.Sh PORTABILITY
.
.Nm
was written in ANSI/C so it should compile on every unix-like platform
supporting an ANSI/C compiler and libvorbis. But this was written on a 
.Fx
system and has not been tested on other platforms. 
.Pp
The supplied Makefile is a BSD-style Makefile and uses the 
.Fx
infrastructure. This is not portable, but as 
.Nm
is pretty simple there should be no problem to write a simple Makefile for
other platforms.
.
.
.Sh DIAGNOSTICS
.
Messages should be self-explanatory.
.Pp
.Ex -std
.
.
.Sh SEE ALSO
.
.Xr ogg123 1 ,
.Xr environ 7
.
.
.Sh AUTHORS
.
.An "Tobias Rehbein" Aq tobias.rehbein@web.de
.
.
.Sh BUGS
.
.Nm
doesn't explicitly support wide or multibyte characters.
.Pp
If you find a bug drop me a mail or file an issue at 
.Dl http://github.com/tobreh/oggquiz/issues
.
.


oggquiz.c
/*-
 * "THE BEER-WARE LICENSE" (Revision 42):
 * <tobias.rehbein@web.de> wrote this file. As long as you retain this notice
 * you can do whatever you want with this stuff. If we meet some day, and you
 * think this stuff is worth it, you can buy me a beer in return.
 *                                                              Tobias Rehbein
 */

#include <assert.h>
#include <err.h>
#include <locale.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <time.h>

#include "options.h"
#include "oggfile.h"
#include "player.h"
#include "ui.h"

#define MIN(a, b)                       ((a) < (b)) ? (a) : (b)

enum {
        FILENAMELEN = 1024
};

static void     avoid_zombie_processes(void);
static int      new_turn(struct ogg_oggfile *oggfiles, struct plr_context *plr_ctx, struct ui_context *ui_ctx, struct opts_options *opts);

int
main(int argc, char **argv)
{
        char            filename[FILENAMELEN];
        char           *newline;
        int             oggfileno;
        struct ogg_oggfile *oggfiles;
        struct opts_options opts;
        struct ogg_context *ogg_ctx;
        struct plr_context *plr_ctx;
        struct ui_context *ui_ctx;

        assert(argc >= 0);
        assert(argv != NULL);

        if (setlocale(LC_ALL, "") == NULL)
                warnx("could not set locale");

        opts_parse_options(&opts, argc, argv);
        /*
         * After this point the opts_options structure is considered read
         * only!
         */

        if ((ogg_ctx = ogg_context_open()) == NULL)
                errx(EX_SOFTWARE, "could not open oggfile context");
        if ((plr_ctx = plr_context_open(opts.ogg123, opts.ogg123_options)) == NULL)
                errx(EX_SOFTWARE, "could not open player context");
        if ((ui_ctx = ui_context_open()) == NULL)
                errx(EX_SOFTWARE, "could not open ui context");
        if ((oggfiles = malloc(opts.choices * sizeof(struct ogg_oggfile))) == NULL)
                err(EX_SOFTWARE, "could not malloc oggfiles");
        avoid_zombie_processes();

        oggfileno = 0;
        while (fgets(filename, FILENAMELEN, stdin) != NULL) {
                if ((newline = strchr(filename, '\n')) != NULL)
                        newline[0] = '\0';

                if (ogg_oggfile_create(ogg_ctx, &oggfiles[oggfileno], filename) == 0)
                        oggfileno++;

                if (oggfileno == opts.choices) {
                        oggfileno = 0;
                        if (new_turn(oggfiles, plr_ctx, ui_ctx, &opts))
                                break;
                }
        }

        plr_stop(plr_ctx);
        plr_context_close(plr_ctx);
        ui_context_close(ui_ctx);
        if (ogg_context_close(ogg_ctx) != 0)
                warn("could not close oggfile context");

        return (0);
}

static int
new_turn(struct ogg_oggfile *oggfiles, struct plr_context *plr_ctx, struct ui_context *ui_ctx, struct opts_options *opts)
{
        static int      first_invocation = 0;
        static struct ui_model model;
        int             correct;
        int             i;
        char            guess;
        time_t          start;

        assert(oggfiles != NULL);
        assert(opts != NULL);

        if (!first_invocation) {
                first_invocation = 1;
                srand(time(NULL));
                if ((model.scores = malloc(opts->players * sizeof(int))) == NULL)
                        err(EX_SOFTWARE, "could not malloc scores");
                if ((model.oggfiles = malloc(opts->choices * sizeof(struct ui_oggfile))) == NULL)
                        err(EX_SOFTWARE, "could not malloc oggfiles");
                for (i = 0; i < opts->players; i++) {
                        model.scores[i] = 0;
                }
                model.turn = 0;
                model.players = opts->players;
                model.choices = opts->choices;
        }
        model.turn++;
        model.current_player = ((model.turn - 1) % model.players);
        for (i = 0; i < model.choices; i++) {
                model.oggfiles[i].title = oggfiles[i].title;
                model.oggfiles[i].artist = oggfiles[i].artist;
                model.oggfiles[i].album = oggfiles[i].album;
        }
        correct = rand() % model.choices;
        model.correct = &model.oggfiles[correct];
        ui_display_quiz(ui_ctx, &model);
        plr_play(plr_ctx, oggfiles[correct].filename);
        start = time(NULL);
        do {
                guess = ui_get_key();
                if (guess == 'q')
                        return (1);
        } while (guess < '1' || guess > '0' + opts->choices);
        model.guess = &model.oggfiles[guess - '1'];
        if (model.guess == model.correct)
                model.score_delta = MIN(opts->time, time(NULL) - start);
        else
                model.score_delta = opts->time;
        model.scores[model.current_player] += model.score_delta;
        ui_display_result(ui_ctx, &model);
        if (ui_get_key() == 'q')
                return (1);

        return (0);
}

static void
avoid_zombie_processes()
{
        struct sigaction sa;

        sa.sa_handler = SIG_IGN;
        sa.sa_flags = 0;
        sigemptyset(&(sa.sa_mask));
        sigaddset(&(sa.sa_mask), SIGCHLD);

        sigaction(SIGCHLD, &sa, NULL);
}
